crates/sc-proof-of-time/src/state_manager.rs

```diff
@@ -235,6 +289,107 @@ impl PotChain {
     }
 }
 
+/// Entropy injection state.
+struct EntropyInjection {
+    /// Entropy injection waiting for the target block
+    /// to be imported, so that the entropy can be applied.
+    /// Contains the block hash and the slot number of the block.
+    pending: BTreeMap<BlockNumber, (BlockHash, SlotNumber)>,
+
+    /// Entropy to apply, indexed by slot number.
+    ready: BTreeMap<SlotNumber, BlockHash>,
+
+    /// Expected number of slots per block.
+    slots_per_block: SlotNumber,
+
+    /// Max entropy entries to keep.
+    max_entries: usize,
+}
+
+impl EntropyInjection {
+    /// Creates the entropy state.
+    fn new(slots_per_block: SlotNumber, max_entries: usize) -> Self {
+        Self {
+            pending: BTreeMap::new(),
+            ready: BTreeMap::new(),
+            slots_per_block,
+            max_entries,
+        }
+    }
+
+    /// Handles a block import.
+    fn handle_block_import(
+        &mut self,
+        block_number: BlockNumber,
+        block_hash: BlockHash,
+        slot_number: SlotNumber,
+    ) {
+        self.add_entropy(block_number);
+        self.update_pending(block_number, block_hash, slot_number);
+    }
+
+    /// Updates the entropy ready to be injected.
+    fn add_entropy(&mut self, block_number: BlockNumber) {
+        if (block_number % POT_ENTROPY_INJECTION_INTERVAL) != 0
+            || block_number <= POT_ENTROPY_INJECTION_LOOK_BACK_DEPTH
+        {
+            return;
+        }
+
+        let source_block = block_number - POT_ENTROPY_INJECTION_LOOK_BACK_DEPTH;
+        let (source_block_hash, source_block_slot) = match self.pending.get(&source_block) {
+            Some((block_hash, slot_number)) => (*block_hash, *slot_number),
+            None => {
+                // This could be because of reorg: we already received an import
+                // notification for the same block number, and the pending
+                // entry was moved to ready state.
+                debug!(%source_block, "Adding entropy: missing source block: ");
+                return;
+            }
+        };
```

Comment:
You will have 0 blocks in `pending` after node restart and will silently skip PoT update (only debug level message) that will result in immediate forking of PoT comparing to those who didn't restart.

The whole logic should be inverted: you should not know about block duration in slots or anything like that, you should pull relevant block header from the actual chain at the moment when it is needed.